Check for integer overflow in Bytes.extend

yallop opened the pull request:
Current behaviour:

```ocaml
# Bytes.extend "abc" max_int max_int;;
- : bytes = "\000"
```

With this fix:

```ocaml
# Bytes.extend "abc" max_int max_int;;
Exception: Invalid_argument "Bytes.extend".
```

yallop committed:
Check for overflow in Bytes.extend.

yallop committed:
Add a changelog entry for Bytes.extend overflow check.

gasche commented on Changes:
```diff
@@ -128,6 +128,9 @@ Next version (4.05.0):
 ### Bug fixes
 
+- GPR#934: check for integer overflow in Bytes.extend
+  (Jeremy Yallop)
```
Sorting entries by MPR first, GPR second, this should go as the very last item of the category.

gasche commented on stdlib/bytes.ml:
```diff
@@ -67,7 +67,10 @@ let sub s ofs len =
+  let len = length s + left in
+  if len < 0 then invalid_arg "Bytes.extend" else
+  let len = len + right in
+  if len < 0 then invalid_arg "Bytes.extend" else
```
I find the shadowing used here not very readable. You could name the first variable `len_left`, or declare `let (++) a b = (* check for overflow *) let c = a + b in if c < 0 then invalid_arg "Bytes.extend" else c` and `let len = length s ++ left ++ right`.

yallop commented:
Thanks for the comments, @gasche.  I've updated the patch.

I think this may need a bit more care before merging, though, since it rejects
```ocaml
Bytes.extend "abc" (-4) 4
```
which was previously accepted, and had reasonable behaviour.  I'll come back to this shortly.

yallop committed:
Add tests for Bytes.extend.

yallop commented:
I've updated the overflow check to take care with negative inputs (31f8776) and added a testsuite for `Bytes.extend` that includes tests both for existing behaviour and for the overflow check (5de7108).

gasche commented:
I'm partial to patches that come with testsuite support. Merged.
